### ISSUE TITLE
"Error: Some variables escaped" only happens on js_of_ocaml

### ISSUE THREAD

**Issue by murilogiacometti** [ocsigen/js_of_ocaml#829]
### Compilation

```
ocamlfind ocamlc -w -5-26-20 -package js_of_ocaml -package js_of_ocaml.ppx -linkpkg -o temp.byte error.ml && js_of_ocaml temp.byte
```

### Output

```
js_of_ocaml: Error: Some variables escaped (#1)
```

### Other compilers

Bucklescript and the OCaml bytecode compiler can compile it without errors:

```
bsc -w -5-26-20 error.ml
```

```
ocamlc -w -5-26-20 error.ml
```

### Versions

```
environment     ocaml-base-compiler.4.08.0
ocamlc          4.08.0
js_of_ocaml     3.4.0
bucklescript    5.0.6 (using OCaml4.02.3+BS)
```

### Program (error.ml)
```
;;
try
  let j =
    try try (if true then fun v -> () else fun e -> ()) 0 with s -> raise s with
    | p ->
        ()
  in
  fun s -> 0
with
| o ->
    raise (try try raise Not_found with x -> x with i -> i)
```

**Comment by murilogiacometti** [comment]
It seems to be related to this code: https://github.com/ocsigen/js_of_ocaml/blob/master/compiler/lib/js_assign.ml

**Comment by hhugo** [comment]
How do you generate theses examples :)

**Comment by murilogiacometti** [comment]
I extended the Efftester tool (https://github.com/jmid/efftester) and its type system to support exception raising and handling :) I also changed it to compare Bucklescript and js_of_ocaml because the native and the bytecode compiler have been stable against random programs.

I'll make it available soon! I'll send you the link when I push it to the public repo.

**Comment by gasche** [comment]
@murilogiacometti I have an intern (see [internship topic](http://gallium.inria.fr/~scherer/tmp/outreachy-project.html) working on extending efftester, our code is at https://github.com/gasche/efftester/ . Could you send me an email ( my email is on my profile page ) to discuss/synchronize? Thanks!

(Apologies for not writing directly, but I couldn't find your email from your github profile page.)

**Comment by hhugo** [comment]
The error doesn't trigger when the ml is compiled with debug info.
```
$ ocamlc -g -w -5-26-20 -o temp.byte error.ml
$ _build/default/compiler/js_of_ocaml.exe error.cmo
```

**Comment by hhugo** [comment]
```
ocamlc -g -c -w -5-26-20 error.ml -o with-debug.cmo
ocamlc -c -w -5-26-20 error.ml -o no-debug.cmo
_build/default/compiler/js_of_ocaml.exe with-debug.cmo -o with-debug.js --pretty --disable shortvar --disable deadcode
_build/default/compiler/js_of_ocaml.exe no-debug.cmo   -o no-debug.js   --pretty --disable shortvar --disable deadcode
```

Generated js with debug info
```javascript
(function(joo_global_object)
   {"use strict";
    var
     runtime=joo_global_object.jsoo_runtime,
     global_data=runtime.caml_get_global_data(),
     Stdlib=global_data.Stdlib;
    try
     {try
       {try {1;var _a_=0;(function(v){return 0}(_a_))}catch(s){throw s}}
      catch(_c_){0}
      (function(s){return 0})}
    catch(_b_)
     {try {try {throw Stdlib[8]}catch(x){var i=x}}catch(i$0){var i=i$0}
      throw i}
    var A=[0];
    runtime.caml_register_global(1,A,"A");
    0;
    return}
  (function(){return this}()));
```
Generated js without debug info (`_a_` has the wrong scope)
```javascript
(function(joo_global_object)
   {"use strict";
    var
     runtime=joo_global_object.jsoo_runtime,
     global_data=runtime.caml_get_global_data(),
     Stdlib=global_data.Stdlib;
    try
     {try
       {try {1;var _b_=0;(function(_h_){return 0}(_b_))}catch(_g_){throw _g_}}
      catch(_f_){0}
      (function(_e_){return 0})}
    catch(_c_)
     {try {try {throw Stdlib[8]}catch(_a_){}}catch(_d_){throw _d_}throw _a_}
                                                                        ^^^
    var B=[0];
    runtime.caml_register_global(1,B,"B");
    0;
    return}
  (function(){return this}()));
```

**Comment by murilogiacometti** [comment]
@gasche No worries about writing here directly! I sent you an email.

**Comment by hhugo** [comment]
I minimized the example: 
```
raise (try try raise Not_found with x -> x with i -> i)
```

**Comment by trefis** [comment]
So:
```
$ cat foo.ml
raise (try try raise Not_found with x -> x with i -> i)
```
```
$ ocamlc -c -dlambda foo.ml
(setglobal Foo!
  (seq
    (raise
      (try (try (raise (field 7 (global Stdlib!))) with x/80 x/80) with i/81
        i/81))
    (makeblock 0)))
$ dumpobj ./foo.cmo
## start of ocaml dump of "./foo.cmo"
       0  PUSHTRAP 13
       2  PUSHTRAP 8
       4  GETGLOBALFIELD Stdlib, 7
       7  RAISE
       8  PUSHACC0
       9  POP 1
      11  POPTRAP
      12  RAISE
      13  PUSHACC0
      14  RAISE
      15  SETGLOBAL Foo
## end of ocaml dump of "./foo.cmo"
```
And:
```
$ ocamlc -c -g -dlambda foo.ml
(setglobal Foo!
  (seq
    (raise
      (after foo.ml(1):6-55
        (try
          (try (raise (after foo.ml(1):21-30 (field 7 (global Stdlib!))))
           with x/80 (before foo.ml(1):41-42 x/80))
         with i/81 (before foo.ml(1):53-54 i/81))))
    (pseudo _none_(0)<ghost>:-1--1 (makeblock 0))))
$ dumpobj ./foo.cmo
## start of ocaml dump of "./foo.cmo"
       0  PUSHTRAP 14
       2  PUSHTRAP 8
       4  GETGLOBALFIELD Stdlib, 7
File "foo.ml", line 1, characters 21-30:
       7  RAISE
       8  PUSHACC0
File "foo.ml", line 1, characters 41-42:
       9  POP 1
      11  POPTRAP
      12  BRANCH 17
      14  PUSHACC0
File "foo.ml", line 1, characters 53-54:
      15  POP 1
File "foo.ml", line 1, characters 6-55:
      17  RAISE
      18  SETGLOBAL Foo
## end of ocaml dump of "./foo.cmo"
```

`dumpobj` is in the tools/ directory of the ocaml distribution, but I'm not sure it's installed.

**Comment by hhugo** [comment]
fixed in #830
